src/htbl.ml

```diff
@@ -0,0 +1,354 @@
+(*
+ * Copyright (c) 2023, Carine Morel <carine.morel.pro@gmail.com>
+ *)
+
+module Type = struct
+  type 'b kind = Dummy | Regular of 'b
+  type key = int
+end
+
+module Llist = struct
+  include Type
+
+  type 'a marked = Last | LRemove | Normal of 'a node | Remove of 'a node
+  and 'a node = { key : key; value : 'a kind; next : 'a marked Atomic.t }
+
+  type 'a t = 'a marked Atomic.t
+
+  type 'a local = {
+    prev : 'a marked Atomic.t;
+    curr : 'a marked;
+    next : 'a marked;
+  }
+
+  let init () : 'a t = Atomic.make Last
+
+  let convert_to_normal = function
+    | (Last | Normal _) as node -> node
+    | LRemove -> Last
+    | Remove node -> Normal node
+
+  let mark_to_be_removed = function
+    | Last -> LRemove
+    | Normal node -> Remove node
+    | _ as x -> x
+
+  let rec find_loop key t prev curr =
+    match curr with
+    | Last | LRemove -> (false, { prev; curr; next = Last })
+    | Normal node | Remove node -> (
+        if Atomic.get prev != curr then try_again key t
+        else
+          let next = Atomic.get node.next in
+          match next with
+          | Normal _ | Last ->
+              if node.key >= key then (node.key = key, { prev; curr; next })
+              else find_loop key t node.next next
+          | _ ->
+              let next = convert_to_normal next in
+              if Atomic.compare_and_set prev curr next then
+                find_loop key t prev next
+              else try_again key t)
```

Comment:
If we fail here, then another thread must have succeeded at doing the deletion and, seemingly, we can just continue. Do we need to retry here?

Reply:
Intuitively, I'm not loving this because its likely to generate extra contention when the structure is already busy. But as above, it's also done in the pragmatic linked list paper.